### PR TITLE
Magic extended attributes rework + add 'chunk_list' magic xattr

### PR DESCRIPTION
**Pull request by janpriessnitz**

### CODE REVIEW

**Comment by janpriessnitz** [comment]
@cernvm-bot cloudtest

**Comment by janpriessnitz** [comment]
@jblomer this one is ready for your review.

**Comment by jblomer** [comment]
@cernvm-bot cloudtest

File: cvmfs/magic_xattr.h [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+  std::map<std::string, BaseMagicXattr *> base_xattrs_;
+  std::map<std::string, BaseMagicXattr *> withhash_xattrs_;
+  std::map<std::string, BaseMagicXattr *> regular_xattrs_;
+  std::map<std::string, BaseMagicXattr *> symlink_xattrs_;
+  std::map<std::string, BaseMagicXattr *> authz_xattrs_;
+  MountPoint *mount_point_;
+};
+
+class AuthZMagicXattr : public BaseMagicXattr {
+  virtual bool PrepareValueFenced();
```
Comment (jblomer) [inline cvmfs/magic_xattr.h:52]: Doesn't this render the two virtual methods private, i.e. don't we need the `public` keyword?

File: cvmfs/mountpoint.cc [hunk 2]
```diff
@@ -1167,6 +1167,8 @@ MountPoint *MountPoint::Create(
   mountpoint->CreateTables();
   mountpoint->SetupBehavior();
 
+  mountpoint->magic_xattr_mgr_ = new MagicXattrManager(mountpoint.weak_ref());
```
Comment (jblomer) [inline cvmfs/mountpoint.cc:1170]: Let's put that in `SetupBehavior`.  The `hide_magic_xattrs_` flag can be incorporated in the `MagicXattrManager`

File: cvmfs/magic_xattr.cc [hunk 3]
```diff
@@ -0,0 +1,569 @@
...
+#include <vector>
+
+#include "catalog_mgr_client.h"
+#include "fetch.h"
+#include "logging.h"
+#include "mountpoint.h"
+#include "quota.h"
+#include "signature.h"
+
+#include "magic_xattr.h"
```
Comment (jblomer) [inline cvmfs/magic_xattr.cc:16]: Move up as the first include

File: cvmfs/magic_xattr.cc [hunk 4]
```diff
@@ -0,0 +1,569 @@
...
+  symlink_xattrs_["xfsroot.rawlink"] = symlink_xattrs_["user.rawlink"];
+
+  authz_xattrs_["user.authz"] = new AuthZMagicXattr();
+}
+
+std::string MagicXattrManager::GetListString(catalog::DirectoryEntry *dirent) {
+  std::string result;
+  std::map<std::string, BaseMagicXattr *>::iterator it = base_xattrs_.begin();
+  for (; it != base_xattrs_.end(); ++it) {
+    result += (*it).first + std::string("\0", 1);
```
Comment (jblomer) [inline cvmfs/magic_xattr.cc:66]: Perhaps better `result.push_back('\0')` to avoid one more string construction.

File: cvmfs/magic_xattr.cc [hunk 5]
```diff
@@ -0,0 +1,569 @@
...
+    result = regular_xattrs_[name];
+  } else if (symlink_xattrs_.count(name) > 0) {
+    result = symlink_xattrs_[name];
+  } else if (authz_xattrs_.count(name) > 0) {
+    result = authz_xattrs_[name];
+  } else {
+    return NULL;
+  }
+
+  result->path_ = path;
```
Comment (jblomer) [inline cvmfs/magic_xattr.cc:114]: Hm, here we have a thread-safety issue.  For the mountpoint it's easy, it is fixed for the lifetime of the manager, so in fact the magic xattr objects can get the mountoint on construction.

Path and dirent need to be locked somehow.  I think the magic xattr base class can have `Lock(path, dirent)` and `Release()` methods that make sure that path and dirent stay stable for `PrepareValueFenced()` followed by `Get()`.

The `MagicXattrManager::Get()` can then return an RAII wrapper around the xattr object that calls `Release()` on destruction.

File: cvmfs/magic_xattr.cc [hunk 6]
```diff
@@ -0,0 +1,569 @@
...
+  result->dirent_ = d;
+  result->mount_point_ = mount_point_;
+  return result;
+}
+
+void MagicXattrManager::Register(const std::string &name,
+                                 BaseMagicXattr *magic_xattr)
+{
+  base_xattrs_[name] = magic_xattr;
```
Comment (jblomer) [inline cvmfs/magic_xattr.cc:123]: We should make sure the name does not exist yet.

File: cvmfs/magic_xattr.h (hunk 1 above)
Comment (janpriessnitz) [reply cvmfs/magic_xattr.h:52]: Good question! I haven't thought of that. I compared the magic xattr outputs between the old and the new version and it matched. So I guess it works as expected and override functions inherit the visibility as well.

**Comment by janpriessnitz** [comment]
@jblomer I've addressed your comments. I have cleaned up the commit sequence, but if you want to see what changed, compare this branch against janpriessnitz/magic-xattr-backup branch.

**Comment by jblomer** [comment]
@cernvm-bot cloudtest

File: cvmfs/magic_xattr.h [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+};
+
+class PubkeysMagicXattr : public BaseMagicXattr {
+  std::string pubkeys_;
+
+  virtual bool PrepareValueFenced();
+  virtual std::string GetValue();
+};
+
+class RawlinkMagicXattr : public BaseMagicXattr {
```
Comment (jblomer) [inline cvmfs/magic_xattr.h:246]: For consistency, this should inherit from a `SymlinkMagicXattr` class, i.e. the class hierarchy should mirror the flavor enum.

File: cvmfs/magic_xattr.h [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+ *    (used in "cvmfs.cc")
+ * 3. Get an attribute by name. Specifically, get a RAII wrapper around
+ *    a singleton attribute instance. This means that the attribute instance
+ *    registered with the manager does not get cloned or copied inside Get().
+ *    Instead, member variables are set and the original instance is returned.
+ *    A mutex prevents from race conditions in case of concurrent access.
+ */
+class MagicXattrManager {
+ public:
+  explicit MagicXattrManager(MountPoint *mountpoint, bool hide_magic_xattrs);
```
Comment (jblomer) [inline cvmfs/magic_xattr.h:123]: With two arguments, we can drop the `explicit`

File: cvmfs/magic_xattr.h [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+
+#include "catalog_counters.h"
+#include "directory_entry.h"
+#include "shortstring.h"
+#include "util/exception.h"
+#include "util/string.h"
+
+class MountPoint;
+
+enum MagicXattrFlavor {
```
Comment (jblomer) [inline cvmfs/magic_xattr.h:19]: Let's add a comment explaining the relationship of this enum with the MagicXattr class hierarchy

File: cvmfs/magic_xattr.h [hunk 10]
```diff
@@ -0,0 +1,324 @@
...
+class MagicXattrRAIIWrapper {
+ public:
+  inline MagicXattrRAIIWrapper() : ref_(NULL) { }
+  inline explicit MagicXattrRAIIWrapper(BaseMagicXattr *ref) : ref_(ref) { }
+  inline ~MagicXattrRAIIWrapper() { if (ref_ != NULL) ref_->Release(); }
+  inline BaseMagicXattr* operator->() const { return ref_; }
+  inline bool IsNull() const { return ref_ == NULL; }
+
+ protected:
+  BaseMagicXattr *ref_;
```
Comment (jblomer) [inline cvmfs/magic_xattr.h:96]: Perhaps better named `ptr_`?

File: cvmfs/cvmfs.cc [hunk 11]
```diff
@@ -1880,6 +1609,70 @@ string *g_boot_error = NULL;
...
+      return "never (fixed root catalog)";
+    } else {
+      time_t now = time(NULL);
+      return StringifyInt( (catalogs_valid_until_ - now) / 60);
+    }
+  }
+};
+
+class InodeMaxMagicXattr : public BaseMagicXattr {
+  virtual bool PrepareValueFenced() { return true; }
```
Comment (jblomer) [inline cvmfs/cvmfs.cc:1635]: Can be dropped.

File: cvmfs/cvmfs.cc [hunk 12]
```diff
@@ -1880,6 +1609,70 @@ string *g_boot_error = NULL;
...
+  virtual std::string GetValue() {
+    return StringifyInt(
+      cvmfs::inode_generation_info_.inode_generation +
+      mount_point_->catalog_mgr()->inode_gauge());
+  }
+};
+
+class MaxFdMagicXattr : public BaseMagicXattr {
+  virtual bool PrepareValueFenced() { return true; }
```
Comment (jblomer) [inline cvmfs/cvmfs.cc:1644]: Can be dropped

File: cvmfs/cvmfs.cc [hunk 13]
```diff
@@ -1880,6 +1609,70 @@ string *g_boot_error = NULL;
...
+  virtual std::string GetValue() {
+    return StringifyInt(cvmfs::max_open_files_ - cvmfs::kNumReservedFd);
+  }
+};
+
+class PidMagicXattr : public BaseMagicXattr {
+  virtual bool PrepareValueFenced() { return true; }
```
Comment (jblomer) [inline cvmfs/cvmfs.cc:1651]: Can be dropped

File: cvmfs/cvmfs.cc [hunk 14]
```diff
@@ -1880,6 +1609,70 @@ string *g_boot_error = NULL;
...
+  virtual std::string GetValue() { return StringifyInt(cvmfs::pid_); }
+};
+
+class UptimeMagicXattr : public BaseMagicXattr {
+  virtual bool PrepareValueFenced() { return true; }
```
Comment (jblomer) [inline cvmfs/cvmfs.cc:1656]: Can be dropped

File: cvmfs/magic_xattr.cc [hunk 15]
```diff
@@ -0,0 +1,515 @@
...
+  if (hide_magic_xattrs()) {
+    return "";
+  }
+
+  std::string result;
+  std::map<std::string, BaseMagicXattr *>::iterator it = xattr_list_.begin();
+  for (; it != xattr_list_.end(); ++it) {
+    MagicXattrFlavor flavor = (*it).second->GetXattrFlavor();
+    // Skip those which should not be displayed
+    switch (flavor) {
```
Comment (jblomer) [inline cvmfs/magic_xattr.cc:74]: Let's add a default which aborts to make sure we don't forget any flavor.

File: cvmfs/magic_xattr.cc [hunk 16]
```diff
@@ -0,0 +1,515 @@
...
+                                             catalog::DirectoryEntry *d)
+{
+  BaseMagicXattr *result;
+  if (xattr_list_.count(name) > 0) {
+    result = xattr_list_[name];
+  } else {
+    return MagicXattrRAIIWrapper(NULL);
+  }
+
+  result->mount_point_ = mount_point_;
```
Comment (jblomer) [inline cvmfs/magic_xattr.cc:108]: We can set this on construction / in `Register()`

File: cvmfs/magic_xattr.cc [hunk 17]
```diff
@@ -0,0 +1,515 @@
...
+#include "mountpoint.h"
+#include "quota.h"
+#include "signature.h"
+
+MagicXattrManager::MagicXattrManager(MountPoint *mountpoint,
+                                     bool hide_magic_xattrs)
+  : mount_point_(mountpoint),
+    hide_magic_xattrs_(hide_magic_xattrs)
+{
+  xattr_list_["user.catalog_counters"] = new CatalogCountersMagicXattr();
```
Comment (jblomer) [inline cvmfs/magic_xattr.cc:23]: Let's use our own `Register()` (e.g. to set the mountpoint)

File: cvmfs/magic_xattr.h [hunk 18]
```diff
@@ -0,0 +1,324 @@
...
+ * 1. Register additional attributes
+ * 2. Get a string containing zero-byte delimited list of attribute names
+ *    (used in "cvmfs.cc")
+ * 3. Get an attribute by name. Specifically, get a RAII wrapper around
+ *    a singleton attribute instance. This means that the attribute instance
+ *    registered with the manager does not get cloned or copied inside Get().
+ *    Instead, member variables are set and the original instance is returned.
+ *    A mutex prevents from race conditions in case of concurrent access.
+ */
+class MagicXattrManager {
```
Comment (jblomer) [inline cvmfs/magic_xattr.h:121]: That should inherit from `SingleCopy` because it managed the magic xattr objects.

File: cvmfs/magic_xattr.h [hunk 19]
```diff
@@ -0,0 +1,324 @@
...
+};
+
+/**
+ * This wrapper ensures that the attribute instance "ref_" is
+ * released after the user finishes using it (on wrapper destruction).
+ */
+class MagicXattrRAIIWrapper {
+ public:
+  inline MagicXattrRAIIWrapper() : ref_(NULL) { }
+  inline explicit MagicXattrRAIIWrapper(BaseMagicXattr *ref) : ref_(ref) { }
```
Comment (jblomer) [inline cvmfs/magic_xattr.h:90]: I think for symmetry reasons, we should call `Lock()` in the RAII constructor

File: cvmfs/magic_xattr.cc [hunk 20]
```diff
@@ -0,0 +1,515 @@
...
+  if (host_chain.size()) {
+    return std::string(host_chain[current_host]);
+  } else {
+    return "internal error: no hosts defined";
+  }
+}
+
+std::string ExternalTimeoutMagicXattr::GetValue() {
+  unsigned seconds, seconds_direct;
+  mount_point_->download_mgr()->GetTimeout(&seconds, &seconds_direct);
```
Comment (jblomer) [inline cvmfs/magic_xattr.cc:240]: This needs to be the external download manager

File: cvmfs/magic_xattr.cc [hunk 21]
```diff
@@ -0,0 +1,515 @@
...
+std::string NDirOpenMagicXattr::GetValue() {
+  return mount_point_->file_system()->n_fs_dir_open()->ToString();
+}
+
+std::string NDownloadMagicXattr::GetValue() {
+  return  mount_point_->statistics()->Lookup("fetch.n_downloads")->Print();
+}
+
+bool NIOErrMagicXattr::PrepareValueFenced() {
+  n_io_err_ = mount_point_->file_system()->n_io_error()->ToString();
```
Comment (jblomer) [inline cvmfs/magic_xattr.cc:343]: We can move this to `GetValue()`, can't we?

File: cvmfs/magic_xattr.cc [hunk 22]
```diff
@@ -0,0 +1,515 @@
...
+  n_used_fd_ = mount_point_->file_system()->no_open_files()->ToString();
+  return true;
+}
+
+std::string UsedFdMagicXattr::GetValue() {
+  return n_used_fd_;
+}
+
+bool UsedDirPMagicXattr::PrepareValueFenced() {
+  n_used_dirp_ = mount_point_->file_system()->no_open_dirs()->ToString();
```
Comment (jblomer) [inline cvmfs/magic_xattr.cc:505]: Can be moved to `GetValue`, I think

File: cvmfs/magic_xattr.cc [hunk 23]
```diff
@@ -0,0 +1,515 @@
...
+}
+
+std::string TimeoutDirectMagicXattr::GetValue() {
+  unsigned seconds, seconds_direct;
+  mount_point_->download_mgr()->GetTimeout(&seconds, &seconds_direct);
+  return StringifyUint(seconds_direct);
+}
+
+bool UsedFdMagicXattr::PrepareValueFenced() {
+  n_used_fd_ = mount_point_->file_system()->no_open_files()->ToString();
```
Comment (jblomer) [inline cvmfs/magic_xattr.cc:496]: Can be moved to `GetValue`, I think.

**Comment by jblomer** [comment]
@cernvm-bot cloudtest cc7

**Pull request merged**